valve_grammar.ne

```diff
@@ -1,97 +1,82 @@
 @builtin "string.ne"
 @builtin "whitespace.ne"
 @{%
-  const first = d => d[0]
-  const ffirst = d => d[0][0]
   const flatten = list => list.reduce((a, b) => a.concat(Array.isArray(b) ? flatten(b) : b), [])
-  const objects = list => list.filter(item => item && typeof item == "object")
+  const object = list => list.filter(item => item && typeof item == "object")[0]
   const join = d => flatten(d).join('')
 %}
 
-expression -> negation | disjunction | datatype | function {% id %}
-
-negation -> "not" _ expression {%
-  function(d) {
-    return {
-      type: "negation",
-      expression: d[2][0],
-    } } %}
-
-disjunction -> expression (_ "or" _ expression):+ {%
+expression -> _ ( string | function ) _ {%
   function(d) {
-    return {
-      type: "disjunction",
-      disjuncts: objects(flatten(d)),
-    } } %}
+    return d[1][0]
+    }%}
 
 function -> function_name "(" arguments ")" {%
   function(d) {
-    return [{
+    return {
       type: "function",
-      name: d[0][0],
+      name: d[0],
       args: d[2],
-    }] } %}
+    }}%}
 
-function_name -> WORD
+function_name -> ALPHANUM {% id %}
 
-arguments -> argument ("," _ argument):* {%
+arguments -> _ argument (_ "," _ argument):* _ {%
   function(d) {
-    return flatten(d).filter(item => item && item != ",");
-  } %}
+    return flatten(d).filter(item => item && item != ",")
+  }%}
 
-argument -> datatype | dqstring | field | function | int | named_arg | regex
+argument -> string | field | function | named_arg | regex
 
-field -> label "." label {% function(d) {
-  return {
-    type: "field",
-    table: d[0][0],
-    column: d[2][0],
-  }}%}
+field -> label "." label {%
+  function(d) {
+    return {
+      type: "field",
+      table: d[0],
+      column: d[2],
+    }}%}
+
+named_arg -> label "=" label {%
+  function(d) {
+    return {
+      type: "named_arg",
+      key: d[0],
+      value: d[2],
+    }}%}
 
-datatype -> [A-Za-z] ALPHANUM {%
+string -> label {%
   function(d) {
     return {
-      type: "datatype",
-      name: d[0] + d[1],
-    } } %}
+      type: "string",
```

Comment:
Do we even need to say "type: string"? Previously, the double quoted strings were just returned as a string, not a dict. I think that's simpler, e.g.:
```
args: [ 'd' ]
```
instead of
```
args: [ { type: 'string', value: 'd' } ]
```

Reply:
I'd like to do it this way now, because it's more regular: all expressions and arguments will be dictionaries.